File: framework-learning/nodeWorkspace/chapter01-buffer/demo05.js
/*
* @Author: gaopeng
* @Email:  gaopeng_hdu@163.com
* @Date:   2018-07-26 23:01:30
* @Last Modified by:   gaopeng
* @Last Modified time: 2018-07-26 23:16:12
*/

'use strict';

const buf = Buffer.from([-1, 5]);

console.log(buf.readInt8(0));
// Prints: -1
console.log(buf.readInt8(1));
// Prints: 5
// console.log(buf.readInt8(2));  // Throws ERR_OUT_OF_RANGE


const buf1 = Buffer.from([1, -2]);

console.log(buf1.readUInt8(0));
// Prints: 1
console.log(buf1.readUInt8(1));
// Prints: 254   256 - 2
// console.log(buf.readUInt8(2));  // Throws ERR_OUT_OF_RANGE


const buf2 = Buffer.allocUnsafe(26);

for (let i = 0; i < 26; i++) {
  // 97 是 'a' 的十进制 ASCII 值 
  buf2[i] = i + 97;
}

const buf3 = buf2.slice(0, 3);

// 输出: abc
console.log(buf3.toString('ascii', 0, buf3.length));

buf2[0] = 33;

// 输出: !bc
console.log(buf3.toString('ascii', 0, buf3.length));


const buf4 = Buffer.allocUnsafe(256);

const len = buf4.write('\u00bd + \u00bc = \u00be', 0);

// 输出: 12 个字节: ½ + ¼ = ¾
console.log(`${len} 个字节: ${buf4.toString('utf8', 0, len)}`);

File: framework-learning/pythonWorkspace/chapter04-Functional-Program/demo05.py
# -*- coding: utf-8 -*-
# @Author: gaopeng
# @Date:   2017-01-23 14:34:24
# @Last Modified by:   gaopeng
# @Last Modified time: 2017-01-23 14:51:16

# 偏函数
print(int('12345'))
print(int('12345', base=8))

# 每次利用base会显得非常麻烦，我们可以包装一个函数
def int2(x, base=2):
    return int(x, base)
print(int2('10000000'))

# 利用偏函数
import functools
int2 = functools.partial(int, base=2)
print(int2('10000000'))
